pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/transformer/MetsXml.scala

```diff
@@ -193,62 +220,23 @@ case class MetsXml(root: Elem) {
     *    </mets:fileGrp>
     *  </mets:fileSec>
     *
-    *  For this input we would expect the following output:
-    *
-    *  Map("FILE_0001_OBJECTS" -> FileReference("FILE_0001_OBJECTS", "objects/b30246039_0001.jp2", "image/jp2"),
-    *      "FILE_0002_OBJECTS" -> FileReference("FILE_0002_OBJECTS", "objects/b30246039_0002.jp2", "image/jp2"))
+    *  For the id "FILE_0002_OBJECTS", this function would return:
+    *  FileReference("FILE_0002_OBJECTS", "objects/b30246039_0002.jp2", Some("image/jp2"))
     */
-  private def fileReferences: ListMap[String, FileReference] =
-    (root \ "fileSec" \ "fileGrp")
-      .filterByAttribute("USE", "OBJECTS")
-      .childrenWithTag("file")
-      .map { file =>
-        val key = file \@ "ID"
-        val objectHref = (file \ "FLocat").toList.headOption
-          .map(_ \@ "{http://www.w3.org/1999/xlink}href")
-        val mimeType = Option(file \@ "MIMETYPE").filter(_.nonEmpty)
-        (key, (objectHref, mimeType))
-      }
-      .collect {
-        case (id, (Some(objectHref), mimeType))
-            if id.nonEmpty && objectHref.nonEmpty =>
-          id -> FileReference(id, objectHref, mimeType)
-      } match {
-      case Nil      => ListMap()
-      case nonEmpty => ListMap(nonEmpty.head)
-    }
-
-  /** Valid METS documents should contain a physicalStructMap section, with the
-    *  bottom most divs each representing a physical page, and linking to files
-    *  in the corresponding fileSec structures:
-    *
-    *  <mets:structMap TYPE="PHYSICAL">
-    *    <mets:div DMDID="DMDPHYS_0000" ID="PHYS_0000" TYPE="physSequence">
-    *      <mets:div ADMID="AMD_0001" ID="PHYS_0001" ORDER="1" TYPE="page">
-    *        <mets:fptr FILEID="FILE_0001_OBJECTS" />
-    *        <mets:fptr FILEID="FILE_0001_ALTO" />
-    *      </mets:div>
-    *      <mets:div ADMID="AMD_0002" ID="PHYS_0002" ORDER="2" TYPE="page">
-    *        <mets:fptr FILEID="FILE_0002_OBJECTS" />
-    *        <mets:fptr FILEID="FILE_0002_ALTO" />
-    *      </mets:div>
-    *    </mets:div>
-    *  </mets:structMap>
-    *
-    *  For this input we would expect the following output:
-    *
-    *  Map("PHYS_0001" -> "FILE_0001_OBJECTS",
-    *      "PHYS_0002" -> "FILE_0002_OBJECTS")
-    */
-  private def physicalStructMap: ListMap[String, String] =
-    (root \ "structMap")
-      .filterByAttribute("TYPE", "PHYSICAL")
-      .descendentsWithTag("div")
-      .sortByAttribute("ORDER")
-      .toMapping(
-        keyAttrib = "ID",
-        valueNode = "fptr",
-        valueAttrib = "FILEID"
+  private def getFileReference(id: String) =
+    for {
+      fileGrp <- root \ "fileSec" \ "fileGrp"
+      objects <- fileGrp.find(_ \@ "USE" == "OBJECTS")
+      listedFiles = objects \ "file"
+      file <- listedFiles.find(_ \@ "ID" == id)
+      objectHref <- (file \ "FLocat").headOption
+        .map(_ \@ "{http://www.w3.org/1999/xlink}href")
+      if objectHref.nonEmpty
```

Comment:
Finding this block pretty tricky to get my head around, but probably mainly due to fact I don't like xpath, and it's not like the original was much easier.

Reply:
XML parsing is hell